controllers/wicd_csr_controller.go

```diff
@@ -0,0 +1,167 @@
+		},
+		validator: validator,
+	}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager
+func (r *wicdCSRReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Build controller for CertificateSigningRequests
+	return builder.ControllerManagedBy(mgr).
+		For(&certificatesv1.CertificateSigningRequest{}).
+		Named(WICDCSRController).
+		Complete(r)
+}
+
+// Reconcile processes WICD CertificateSigningRequests
+func (r *wicdCSRReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("CertificateSigningRequest", req.NamespacedName)
+
+	csr := &certificatesv1.CertificateSigningRequest{}
+	if err := r.client.Get(ctx, req.NamespacedName, csr); err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get CertificateSigningRequest: %w", err)
+	}
+
+	if !r.validator.IsCorrectCertificateType(csr) {
+		return reconcile.Result{}, nil // Not a WICD CSR, ignore
+	}
+
+	if !r.isWICDServiceAccount(csr) && !r.isWICDCertificateIdentity(csr) {
+		return reconcile.Result{}, nil
+	}
+
+	if isCSRApproved(csr) || isCSRDenied(csr) {
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.validator.ValidateCSR(ctx, csr); err != nil {
+		log.Error(err, "WICD CSR validation failed, ignoring CSR")
+		return reconcile.Result{}, nil
+	}
+
+	return r.approveCSR(ctx, csr)
+}
+
+// isWICDServiceAccount checks if this CSR is from the WICD service account
+func (r *wicdCSRReconciler) isWICDServiceAccount(csr *certificatesv1.CertificateSigningRequest) bool {
+	expectedUsername := fmt.Sprintf("system:serviceaccount:%s:%s", r.watchNamespace, windows.WicdServiceName)
+	return csr.Spec.Username == expectedUsername
+}
+
+// isWICDCertificateIdentity checks if this CSR is from a WICD certificate-based identity
+func (r *wicdCSRReconciler) isWICDCertificateIdentity(csr *certificatesv1.CertificateSigningRequest) bool {
+	return strings.HasPrefix(csr.Spec.Username, rbac.WICDUserPrefix)
+}
+
+// approveCSR approves the certificate signing request using the proper UpdateApproval API
+func (r *wicdCSRReconciler) approveCSR(ctx context.Context, csr *certificatesv1.CertificateSigningRequest) (reconcile.Result, error) {
+	log := r.log.WithValues("CertificateSigningRequest", csr.Name)
+
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:           certificatesv1.CertificateApproved,
+		Status:         corev1.ConditionTrue,
+		Reason:         "WICDAutoApproved",
+		Message:        "This CSR was approved by the WICD certificate controller",
+		LastUpdateTime: metav1.Now(),
+	})
+
+	if _, err := r.k8sclientset.CertificatesV1().CertificateSigningRequests().UpdateApproval(ctx, csr.Name, csr, metav1.UpdateOptions{}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to approve CSR: %w", err)
+	}
+	log.Info("CSR approved", "CSR", csr.Name)
```

Comment:
Please move the log message out of this function, that way the duplicate 
```
	log := r.log.WithValues("CertificateSigningRequest", csr.Name)
```
can be removed